packages/vscode-extension/src/project/project.ts

```diff
@@ -110,6 +111,33 @@ export class Project implements Disposable, ProjectInterface, DeviceSessionsMana
     );
   }
 
+  onDeviceSessionSelected(sessionId: DeviceId | undefined): void {
+    this.updateProjectState({ selectedSessionId: sessionId });
+  }
+
+  onDeviceSessionChange(state: DeviceSessionState): void {
+    const deviceId = state.deviceInfo.id;
+    assert(deviceId in this.projectState.deviceSessions, "A session must already exist");
+    const newDeviceSessions = { ...this.projectState.deviceSessions, [deviceId]: state };
+    this.updateProjectState({ deviceSessions: newDeviceSessions });
+  }
+
+  onDeviceSessionStarted(state: DeviceSessionState): void {
+    const deviceId = state.deviceInfo.id;
+    const newDeviceSessions = {
+      ...this.projectState.deviceSessions,
+      [deviceId]: state,
+    };
+    this.updateProjectState({ deviceSessions: newDeviceSessions });
+  }
+
+  onDeviceSessionStopped(state: DeviceSessionState): void {
+    const deviceId = state.deviceInfo.id;
+    const newDeviceSessions = { ...this.projectState.deviceSessions };
+    delete newDeviceSessions[deviceId];
+    this.updateProjectState({ deviceSessions: newDeviceSessions });
+  }
```

Comment:
Maybe it'd be worth moving this wholesale to `DeviceSessionsManager` and having it send `{ selectedDeviceId, deviceSessions }` through `updateProjectState` instead.